streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java

```diff
@@ -0,0 +1,186 @@
+                Thread.sleep(200);
+            } catch (Exception ex) {
+                // do nothing
+            }
+            retry--;
+            lock = tryAcquireLock(channel);
+        }
+        if (lock != null) {
+            locks.put(taskId, lock);
+        }
+        return lock != null;
+    }
+
+    /**
+     * Unlock the state directory for the given {@link TaskId}
+     * @param taskId
+     * @throws IOException
+     */
+    public void unlock(final TaskId taskId) throws IOException {
+        final FileLock lock = locks.remove(taskId);
+        if (lock != null) {
+            lock.release();
+            final FileChannel fileChannel = channels.remove(taskId);
+            if (fileChannel != null) {
+                fileChannel.close();
+            }
+        }
+    }
+
+    /**
+     * Remove the directories for any {@link TaskId}s that are no-longer
+     * owned by this {@link StreamThread} and aren't locked by either
+     * another process or another {@link StreamThread}
+     */
+    public void cleanRemovedTasks() {
+        final File[] taskDirs = listTaskDirectories();
+        if (taskDirs == null || taskDirs.length == 0) {
+            return; // nothing to do
+        }
+
+        for (File taskDir : taskDirs) {
+            final String dirName = taskDir.getName();
+            TaskId id = TaskId.parse(dirName);
+            if (!locks.containsKey(id)) {
+                try {
+                    if (lock(id, 0)) {
+                        log.info("Deleting obsolete state directory {} for task {}", dirName, id);
+                        Utils.delete(taskDir);
+                    }
+                } catch (OverlappingFileLockException e) {
+                    // locked by another thread
+                } catch (IOException e) {
+                    log.error("Failed to lock the state directory due to an unexpected exception", e);
+                } finally {
+                    try {
+                        unlock(id);
+                    } catch (IOException e) {
+                        log.error("Failed to release the state directory lock");
+                    }
+                }
+            }
+        }
+
+    }
+
+    /**
+     * List all of the task directories
+     * @return
+     */
+    public File[] listTaskDirectories() {
+        return stateDir.listFiles(new FileFilter() {
+            @Override
+            public boolean accept(final File pathname) {
+                final String name = pathname.getName();
+                return pathname.isDirectory() && name.matches("\\d+_\\d+");
+            }
+        });
+    }
+
+    private FileChannel getFileChannel(final TaskId taskId, final Path lockPath) throws IOException {
+        final FileChannel channel = FileChannel.open(lockPath, StandardOpenOption.CREATE, StandardOpenOption.WRITE);
+        final FileChannel existing = channels.putIfAbsent(taskId, channel);
+        if (existing != null) {
+            channel.close();
```

Comment:
Thanks @ijuma - i meant to change that. It doesn't need to be a `ConcurrentHashMap` either as there is a 1-to-1 mapping from `StreamThread` to `StateDirectory`. Updating now